### ISSUE TITLE
DeviceIOControl Win32 API call returns incorrect result for SYMLINKD folder mounted to docker volume

### ISSUE THREAD

**Issue by minherz** [moby/moby#31089]
**Description**

In Windows container a call to `DeviceIOControl` Win32 API provides incorrect results when called for a folder which is SYMLINKD to the mounted docker volume.

The problem was originally reported to docker/for-win ([num 506](https://github.com/docker/for-win/issues/506)) and then moved into this project.

**Steps to reproduce the issue:**
  1. Run command
```
docker run -it -v c:\mounted_folder --name=test microsoft/dotnet:nanoserver cmd
```
  2. Use an application that calls to `DeviceIOControl` and prints the resulted fields for folder `c:\mounted_folder`. Observe the path of the SubstituteName field.

  3. Create a symbolic link to the local folder:
```
mklink /D c:\System c:\Windows
```
  4. Use the application again for c:\System and observe the path of the SubstituteName field.

**Describe the results you received:**
When above steps are performed in the windows container the step#2 the displayed value of the SubstituteName will be something like `\ContainerMappedDirectories\A07E52BD-3B5B-44EB-8648-7DDB69D26682` starting with just single backslash.
The output for the step#4 will result in the valid output `\??\c:\Windows`. 

**Describe the results you expected:**
The expected result that the behavior is same and similar to non-container execution, meaning that when a link is created for absolute path it starts with `\??\`.

**Additional information you deem important (e.g. issue happens only occasionally):**
I wrote a simple application that receives a path to SYMLINKD object and prints DeviceIOControl output that can be used. Alternatively the method can be called using `powershell.exe`. The app can be found [here](https://slack-files.com/T3NG3BE8L-F46CVAYEN-8247786785)

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 20
Server Version: 1.13.1
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 14393 (14393.693.amd64fre.rs1_release.161220-1747)
Operating System: Windows 10 Enterprise
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 15.91 GiB
Name: miglleonidy
ID: HUE2:4R36:LCWC:LOUK:BG3M:E465:5FG2:7WNC:7FFS:QFJI:YXMC:JOEE
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 22
 System Time: 2017-02-16T18:24:47.7241544+02:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
  - Diagnostic ID: F9F388E9-7AB0-4CE8-BD59-045CF28124A7/2017-02-16_15-44-33
  - Originally issue was reported to minio project as it was discovered due invalid behavior of `golang` code for processing data storage path (see [issue 3742 in minio/minio](https://github.com/minio/minio/issues/3742)).
  - Operating System and version: Windows 10 x64 Enterprise Version 1607 (Build 14393.693)
  - Docker for Windows version: Version 1.13.1-beta42 (10069) in Beta mode

**Comment by cpuguy83** [comment]
ping @johnstep

**Comment by minherz** [comment]
Hello, @cpugy83, @johnstep,
Any news?

**Comment by johnstep** [comment]
The substitute name in `REPARSE_DATA_BUFFER` is an NT object name, not a Win32 file name. The `\ContainerMappedDirectories\...` name is valid because Windows maps directories into containers through that object directory. However, I believe `\??\ContainerMappedDirectories\...` is also valid, since `\??` is a special directory, mapping to the Win32 namespace for the calling logon session, which contains a symbolic link from `\GLOBAL??\ContainerMappedDirectories` to `\ContainerMappedDirectories` within the container.

How is this impacting your scenario? Are you relying on the `\??\` prefix for something? The main problem I see is that file names of the form `\\?\ContainerMappedDirectories\...` are incompatible with some programs, including parts of `cmd.exe`. While some commands work, such as copying files, others do not.

**Comment by minherz** [comment]
@johnstep, it seems that the behavior of docker on linux is different. the minio project implements parsing of the symlink path basing that the prefix starts with `\??\` which stops working when called in docker for windows (see  [issue 3742 in minio/minio](https://github.com/minio/minio/issues/3742)).
If there is indeed discrepancy between docker behavior in linux and windows when a symlink is created it should be either documented or (better) brought to be the same.
Otherwise, like in example you provided, some applications will not be able work equally on linux and windows containers.

**Comment by StefanScherer** [comment]
@minherz Seems that was a bug in golang, the minio project has ported the bugfix of upcoming golang into minio itself: https://github.com/minio/minio/pull/4203 The latest release now should work in a Windows container.

**Comment by minherz** [comment]
@StefanScherer Unfortunately, minio team wrote that the fix depends on upgrade of the minio to next golang version which was postponed due some internal issues. I am honestly do not remember why I was asked to open an issue for this project. It is possible that minio uses moby's modules or something like that.

**Comment by StefanScherer** [comment]
@minherz They have backported the fix for golang 1.9 into minio codebase so they can keep at their current golang version in use. Sure other golang modules using os.Stat() directly still have this problem until next golang version will be used.

**Comment by minherz** [comment]
OK. I close it since I saw the changes for workaround merged.
